### PR TITLE
[Enhancement](show-backends-disks) Add show backends disks

### LINKED ISSUE #24226: [Enhancement] Add statement to query disk information corresponding to data directory of BE node
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Usually a data directory corresponds to a disk, In actual applications, we usually need to query the usage of different disks, and the disk information displayed in ‘show backends’ is the usage of all disks. And if other disks of the BE are full, the stability of BE will be affected.
So it is necessary to add statement to query disk information corresponding to data and log directory of BE node.

### Solution

Add statement to query disk information corresponding to data and log directory of BE node.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by yongjinhou**
…ry of BE node

## Proposed changes

Issue Number: close #24226

Add statement to query disk information corresponding to data and log directory of BE node.

'show backends disks;'
+-----------+-------------+------------------------------+---------+----------+---------------+-------------+-------------------+---------+ 
| BackendId | Host        | RootPath                     | DirType | DiskState| TotalCapacity | UsedCapacity| AvailableCapacity | UsedPct |
+-----------+-------------+------------------------------+---------+----------+---------------+-------------+-------------------+---------+
| 10002     | 10.xx.xx.90 | /home/work/output/be/storage | STORAGE | ONLINE   | 7.049 TB      | 2.478 TB    | 4.571 TB          | 35.16 % |
| 10002     | 10.xx.xx.90 | /home/work/output/be         | DEPLOY  | ONLINE   | 7.049 TB      | 2.478 TB    | 4.571 TB          | 35.16 % |
| 10002     | 10.xx.xx.90 | /home/work/output/be/log     | LOG     | ONLINE   | 7.049 TB      | 2.478 TB    | 4.571 TB          | 35.16 % |
+-----------+-------------+------------------------------+---------+----------+---------------+-------------+-------------------+---------+

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run clickbench

**Comment by yongjinhou** [comment]
run p0

**Comment by yongjinhou** [comment]
run p0

File: docs/zh-CN/docs/sql-manual/sql-reference/Show-Statements/SHOW-BACKENDS-DISKS.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+| BackendId | Host        | RootPath                       | TotalCapacity | DataUsedCapacity | RemoteUsedCapacity | TrashUsedCapacity | DiskUsedCapacity | AvailableCapacity | UsedPct | StorageMedium |
++-----------+-------------+--------------------------------+---------------+------------------+--------------------+-------------------+------------------+-------------------+---------+---------------+
+| 10004     | 10.xx.xx.90 | /home/disk1/storage            | 7.049 TB      | 0.000            | 0.000              | 0.000             | 2.465 TB         | 4.585 TB          | 34.96 % | HDD           |
+| 10005     | 10.xx.xx.91 | /home/disk1/storage            | 7.049 TB      | 0.000            | 0.000              | 0.000             | 1.858 TB         | 5.192 TB          | 26.35 % | HDD           |
++-----------+-------------+--------------------------------+---------------+------------------+--------------------+-------------------+------------------+-------------------+---------+---------------+
+2 rows in set (0.00 sec)
+`
+### Keywords
+
+    SHOW, BACKENDS
```
Comment (luozenglin) [inline docs/zh-CN/docs/sql-manual/sql-reference/Show-Statements/SHOW-BACKENDS-DISKS.md:67]: SHOW, BACKENDS, DISK, DISKS

**Review by luozenglin — DISMISSED** [review]
LGTM

File: docs/zh-CN/docs/sql-manual/sql-reference/Show-Statements/SHOW-BACKENDS-DISKS.md (hunk 1 above)
Comment (yongjinhou) [reply docs/zh-CN/docs/sql-manual/sql-reference/Show-Statements/SHOW-BACKENDS-DISKS.md:67]: I will add it when implementing corresponding tvf.

**Review by xy720 — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/common/proc/BackendsProcDir.java [hunk 2]
```diff
@@ -184,6 +204,80 @@ public static List<List<String>> getBackendInfos() {
...
+        List<List<Comparable>> comparableBackendsDiskInfos = new LinkedList<>();
+        for (long backendId : backendIds) {
+            Backend backend = systemInfoService.getBackend(backendId);
+            if (backend == null) {
+                continue;
+            }
+
+            ImmutableMap<String, DiskInfo> disksRef = backend.getDisks();
+            for (DiskInfo disk : disksRef.values()) {
+                if (disk.getState() == DiskState.ONLINE) {
```
Comment (xy720) [inline fe/fe-core/src/main/java/org/apache/doris/common/proc/BackendsProcDir.java:229]: May be we should also show the offline disk. The offline disk is also reported from be to master. (Offline disk means the disk health check is bad). And we should add a column to show this disk's online/offline state.

Comment (yongjinhou) [reply fe/fe-core/src/main/java/org/apache/doris/common/proc/BackendsProcDir.java:229]: ok

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run p0

**Comment by yongjinhou** [comment]
run p0

**Comment by yongjinhou** [comment]
run p0

**Review by xy720 — APPROVED** [review]
LGTM

**Review by xy720 — CHANGES_REQUESTED** [review]
Log dir should not report as a normal disk. It will be a dest path in Fe TabletScheduler, may influence backend/disk rebalance.

File: be/src/olap/storage_engine.cpp [hunk 3]
```diff
@@ -357,6 +358,15 @@ Status StorageEngine::get_all_data_dir_info(std::vector<DataDirInfo>* data_dir_i
     return res;
 }
 
+Status StorageEngine::get_log_dir_info(LogDirInfo* log_dir_infos) {
+    Status res = Status::OK();
+    _log_dir->health_check();
+    _log_dir->update_capacity();
+    LogDirInfo dir_info = _log_dir->get_dir_info();
```
Comment (luozenglin) [inline be/src/olap/storage_engine.cpp:365]: Why not just pass the log_dir_infos pointer as an argument to get_dir_info()?

File: gensrc/thrift/MasterService.thrift [hunk 4]
```diff
@@ -76,7 +76,7 @@ struct TDisk {
     1: required string root_path
     2: required Types.TSize disk_total_capacity
     // local used capacity
-    3: required Types.TSize data_used_capacity
+    3: optional Types.TSize data_used_capacity
```
Comment (luozenglin) [inline gensrc/thrift/MasterService.thrift:79]: Consider that when upgrading, be applies the new code, but fe is still REQUIRED, which will cause an exception.

**Comment by luozenglin** [comment]
The be execution directory may be full of core files, so displaying disk information about the be execution directory is necessary.

File: docs/zh-CN/docs/sql-manual/sql-reference/Show-Statements/SHOW-BACKENDS-DISKS.md (hunk 1 above)
Comment (yongjinhou) [reply docs/zh-CN/docs/sql-manual/sql-reference/Show-Statements/SHOW-BACKENDS-DISKS.md]: done

File: be/src/agent/task_worker_pool.cpp [hunk 5]
```diff
@@ -678,6 +679,19 @@ void TaskWorkerPool::_report_disk_state_worker_thread_callback() {
             disk.__set_used(root_path_info.is_used);
             request.disks[root_path_info.path] = disk;
         }
+
+        LogDirInfo log_dir_info;
+        StorageEngine::instance()->get_log_dir_info(&log_dir_info);
+        LOG(INFO) << "path: " << log_dir_info.path << " total capacity: " << log_dir_info.capacity
```
[the remark this answers is not in the document]
Comment (yongjinhou) [reply be/src/agent/task_worker_pool.cpp:685]: done

**Comment by yongjinhou** [comment]
> The be execution directory may be full of core files, so displaying disk information about the be execution directory is necessary.

done

File: gensrc/thrift/MasterService.thrift (hunk 4 above)
Comment (yongjinhou) [reply gensrc/thrift/MasterService.thrift]: Set it to required

File: be/src/olap/storage_engine.cpp (hunk 3 above)
Comment (yongjinhou) [reply be/src/olap/storage_engine.cpp]: done

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run feut

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
> Log dir should not report as a normal disk. It will be a dest path in Fe TabletScheduler, may influence backend/disk rebalance.



> Log dir should not report as a normal disk. It will be a dest path in Fe TabletScheduler, may influence backend/disk rebalance.

I categorized the disk types

**Comment by yongjinhou** [comment]
run p0

**Comment by yongjinhou** [comment]
run p0

**Review by xy720 — DISMISSED** [review]
LGTM

File: gensrc/thrift/MasterService.thrift [hunk 6]
```diff
@@ -83,6 +83,7 @@ struct TDisk {
     7: optional Types.TStorageMedium storage_medium
     8: optional Types.TSize remote_used_capacity
     9: optional Types.TSize trash_used_capacity
+    10: optional string dir_type
```
Comment (luozenglin) [inline gensrc/thrift/MasterService.thrift:86]: It is better to definate a enum in Types.thrift

**Review by chenlinzhong — DISMISSED** [review]
LGTE

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run compile

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run feut

**Comment by yongjinhou** [comment]
run buildall

**Comment by yongjinhou** [comment]
run beut

**Review by xy720 — DISMISSED** [review]
LGTM

**Comment by yongjinhou** [comment]
run buildall

File: be/src/agent/task_worker_pool.h [hunk 7]
```diff
@@ -19,6 +19,7 @@
 
 #include <butil/macros.h>
```
[the remark this answers is not in the document]
Comment (yongjinhou) [reply be/src/agent/task_worker_pool.h:20]: 不要在这乱指挥了

**Comment by yongjinhou** [comment]
run p0

**Comment by adonis0147** [comment]
run buildall

**Review by chenlinzhong — APPROVED** [review]
LGTM

**Review by xy720 — APPROVED** [review]
LGTM

**Pull request merged**